### PR TITLE
fix: [CI-5919]: fixed an issue with CI strategy expansion

### PR DESCRIPTION
**Pull request by devkimittal**
## Checklist
- [ ] I've documented the changes in the PR description.
- [ ] I've tested this change either in PR or local environment.
- [ ] If hashcheck failed I followed [the checklist](https://harness.atlassian.net/wiki/spaces/DEL/pages/21016838831/PR+Codebasehash+Check+merge+checklist) and updated this PR description with my answers to make sure I'm not introducing any breaking changes.

## Comment Triggers
<details>
  <summary>Build triggers</summary>

- Feature build: `trigger feature-build`
- Immutable delegate `trigger publish-delegate`
</details>

<details>
  <summary>PR Check triggers</summary>

You can run multiple PR check triggers by comma separating them in a single comment. e.g. `trigger ti0, ti1`

- Compile: `trigger compile`
- CodeformatCheckstyle: `trigger checkstylecodeformat`
    - CodeFormat: `trigger codeformat`
    - Checkstyle: `trigger checkstyle`
- MessageMetadata: `trigger messagecheck`
- Recency: `trigger recency`
- BuildNumberMetadata: `trigger buildnum`
- PMD: `trigger pmd`
- Copyright Check: `trigger copyrightcheck`
- Feature Name Check: `trigger featurenamecheck`
- TI-bootstrap: `trigger ti0`
- TI-bootstrap1: `trigger ti1`
- TI-bootstrap2: `trigger ti2`
- TI-bootstrap3: `trigger ti3`
- TI-bootstrap4: `trigger ti4`
- FunctionalTest1: `trigger ft1`
- FunctionalTest2: `trigger ft2`
- CodeBaseHash: `trigger codebasehash`
- CodeFormatCheckstyle: `trigger checkstylecodeformat`
- SonarScan: `trigger ss`
- Trigger all Checks: `trigger smartchecks`
</details>

## PR check failures and solutions
https://harness.atlassian.net/wiki/spaces/BT/pages/21106884744/PR+Checks+-+Failures+and+Solutions

## [Contributor license agreement](https://github.com/harness/harness-core/blob/develop/CONTRIBUTOR_LICENSE_AGREEMENT.md)

- - -
This change is [Reviewable](https://reviewable.io/reviews/harness/harness-core/40718)

### CODE REVIEW

**Comment by devkimittal** [comment]
trigger messagecheck

File: 332-ci-manager/service/src/main/java/io/harness/ci/execution/buildstate/BuildSetupUtils.java [hunk 1]
```diff
@@ -31,9 +48,13 @@ public class BuildSetupUtils {
   @Inject private VmInitializeTaskParamsBuilder vmInitializeTaskParamsBuilder;
   @Inject private DliteVmInitializeTaskParamsBuilder dliteVmInitializeTaskParamsBuilder;
   @Inject private DockerInitializeTaskParamsBuilder dockerInitializeTaskParamsBuilder;
+  @Inject private CILicenseService ciLicenseService;
+  @Inject private StrategyHelper strategyHelper;
 
   public CIInitializeTaskParams getBuildSetupTaskParams(
       InitializeStepInfo initializeStepInfo, Ambiance ambiance, String logPrefix) {
+    initializeStepInfo = populateStrategyExpansion(initializeStepInfo, ambiance);
```
Comment (raghavharness) [inline 332-ci-manager/service/src/main/java/io/harness/ci/execution/buildstate/BuildSetupUtils.java:56]: let's call this in InitializeTaskStepV2 ?

File: 332-ci-manager/service/src/main/java/io/harness/ci/execution/buildstate/BuildSetupUtils.java [hunk 2]
```diff
@@ -49,4 +70,31 @@ public CIInitializeTaskParams getBuildSetupTaskParams(
     }
     return null;
   }
+
+  private InitializeStepInfo populateStrategyExpansion(InitializeStepInfo initializeStepInfo, Ambiance ambiance) {
```
Comment (raghavharness) [inline 332-ci-manager/service/src/main/java/io/harness/ci/execution/buildstate/BuildSetupUtils.java:74]: why to return InitializeStepInfo ? InitializeStepInfo is passed by reference right so setting the values should be enough.

**Review by raghavharness — CHANGES_REQUESTED** [review]

**Review by raghavharness — DISMISSED** [review]
:lgtm:

> Reviewed 1 of 2 files at r1, 1 of 7 files at r3, 5 of 6 files at r4, 1 of 1 files at r5, all commit messages.
> *[Reviewable](https://reviewable.io/reviews/harness/harness-core/40718#-:-NHxo6KO4zzdJlC8BBIo:bnfp4nl)* status: :shipit: complete! ✓

**Review by brijeshdhakar — CHANGES_REQUESTED** [review]
> Reviewed 1 of 2 files at r1, 1 of 3 files at r2, 1 of 7 files at r3, 5 of 6 files at r4, 1 of 1 files at r5, all commit messages.
> *[Reviewable](https://reviewable.io/reviews/harness/harness-core/40718)* status: :shipit: complete! ✓

___
*[`332-ci-manager/service/src/main/java/io/harness/ci/execution/states/V1/InitializeTaskStepV2.java` line 434 at r5](https://reviewable.io/reviews/harness/harness-core/40718#-NHxofs64wqQ8fzlkdeF:-NHxofs64wqQ8fzlkdeG:bjwlbj3) ([raw file](https://github.com/harness/harness-core/blob/e1fb62e0b246d8666cb4a2c584ded00c1775ff39/332-ci-manager/service/src/main/java/io/harness/ci/execution/states/V1/InitializeTaskStepV2.java#L434)):*
> ```Java
>   }
> 
>   private void populateStrategyExpansion(InitializeStepInfo initializeStepInfo, Ambiance ambiance) {
> ```

How does moving the expansion from generator to task solve the issue?

**Comment by devkimittal** [comment]
*[`332-ci-manager/service/src/main/java/io/harness/ci/execution/states/V1/InitializeTaskStepV2.java` line 434 at r5](https://reviewable.io/reviews/harness/harness-core/40718#-NHxofs64wqQ8fzlkdeF:-NHxpRZJ-5oylj1UHPwM:b-giz2xt) ([raw file](https://github.com/harness/harness-core/blob/e1fb62e0b246d8666cb4a2c584ded00c1775ff39/332-ci-manager/service/src/main/java/io/harness/ci/execution/states/V1/InitializeTaskStepV2.java#L434)):*
<details><summary><i>Previously, brijeshdhakar (Brijesh Dhakar) wrote…</i></summary><blockquote>

How does moving the expansion from generator to task solve the issue?
</blockquote></details>

The generator runs for all the stages before pipeline has actually run, since second stage requires the output of first stage, it will fail.
InitTask however runs before every stage and will have the output from stage 1 since it's execution is completed.

**Review by brijeshdhakar — APPROVED** [review]
> Reviewed 1 of 7 files at r3, 1 of 1 files at r6, all commit messages.
> *[Reviewable](https://reviewable.io/reviews/harness/harness-core/40718)* status: :shipit: complete! ✓

**Review by harshjain12 — APPROVED** [review]
:lgtm:

> Reviewed 1 of 2 files at r1, 1 of 7 files at r3, 1 of 1 files at r6, all commit messages.
> *[Reviewable](https://reviewable.io/reviews/harness/harness-core/40718#-:-NHxzTzN6dza7Rdx5goH:bnfp4nl)* status: :shipit: complete! ✓

**Pull request merged**